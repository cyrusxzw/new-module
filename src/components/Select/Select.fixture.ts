const SelectFixture = {
  className: '',
  dataTestRef: 'SELECT_TEST_REF',
  errorMessage: 'Test Error Messsage',
  label: 'Vestibulum accumsan',
  name: 'select',
  onBlur: () => undefined,
  onChange: () => undefined,
  onFocus: () => undefined,
  options: [
    {
      id: 'morbi-id-nulla',
      label: 'Morbi id nulla',
      value: 'morbi-id-nulla',
    },
    {
      id: 'nullam-ultrices',
      label: 'Nullam ultrices',
      value: 'nullam-ultrices',
    },
  ],
  theme: 'dark',
  value: 'morbi-id-nulla',
};

export { SelectFixture };
