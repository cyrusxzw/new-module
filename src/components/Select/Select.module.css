@import 'styles/index';

.base {
  position: relative;
  display: block;
  padding-top: 25px;
  margin-bottom: 15px;
  font-family: var(--font-family-sans-serif-regular);

  @media (--viewport-md) {
    max-width: 466.66px;
  }

  &,
  & * {
    box-sizing: border-box;
  }
}

.input {
  display: inline-block;
  width: 100%;
  padding: 0;
  border-width: 0 0 1px 0;
  border-style: solid;
  border-color: var(--color-grey-65);
  margin: 0;
  appearance: none;
  background-color: transparent;
  border-radius: 0;
  color: var(--color-dark-copy);
  cursor: pointer;
  font-size: rem(20px);
  line-height: 1.5;
  outline: none;
  text-overflow: ellipsis;

  @media (--viewport-md) {
    font-size: rem(24px);
  }

  .dark & {
    color: var(--color-dark-copy);
  }

  .light & {
    color: var(--color-light-copy);
  }
}

.isBlock {
  padding: 20px 50px 20px 20px;
  border-width: 1px;
  font-size: rem(16px);
  -webkit-padding-end: 50px;
  -webkit-padding-start: 20px;
}

.icon {
  position: absolute;
  top: 35px;
  right: 22px;
  pointer-events: none;

  .dark & {
    fill: var(--color-dark-label);
  }

  .light & {
    fill: var(--color-light-label);
  }
}

.label {
  position: absolute;
  top: 0;
  font-size: rem(20px);
  pointer-events: none;
  transform: translateY(25px);
  transition: all 300ms cubic-bezier(0.215, 0.61, 0.355, 1);

  @media (--viewport-md) {
    font-size: rem(24px);
  }

  .dark & {
    color: var(--color-dark-label);
  }

  .light & {
    color: var(--color-light-label);
  }

  .hasFocus & {
    font-size: rem(14px);
    transform: translateY(0);
  }
}

.errorMessage {
  display: block;
  height: rem(14px);
  margin: 10px 0 0;
  color: var(--color-error);
  font-size: rem(14px);
  line-height: 1.4;
}

.hasError {
  .input {
    border-color: var(--color-error);
  }

  .icon {
    fill: var(--color-error);
  }
}
