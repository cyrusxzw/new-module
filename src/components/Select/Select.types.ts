import type { Themes } from '~/types';

type SelectOption = {
  id?: string;
  label?: string;
  value?: string;
};

type SelectProps = {
  className?: string;
  dataTestRef?: string;
  errorMessage?: string;
  isBlock?: boolean;
  isEnabled?: boolean;
  label?: string;
  name: string;
  onBlur?: (event: React.FocusEvent<HTMLSelectElement>) => void;
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  onFocus?: (event: React.FocusEvent<HTMLSelectElement>) => void;
  options: SelectOption[];
  theme?: Themes;
  value?: string;
};

export type { SelectProps };
