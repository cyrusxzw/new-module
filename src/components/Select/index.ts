import { Select } from './Select';
import type { SelectProps } from './Select.types';

export { Select };
export type { SelectProps };
