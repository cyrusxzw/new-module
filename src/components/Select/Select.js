import React, { useState, forwardRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Icon } from '~/components/Icon';
import styles from './Select.module.css';

const Select = forwardRef(
  (
    {
      className,
      dataTestRef,
      errorMessage,
      isBlock,
      label,
      name,
      onBlur,
      onChange,
      onFocus,
      options,
      theme,
      value,
    },
    ref,
  ) => {
    const [hasFocus, setHasFocus] = useState(false);
    const classSet = cx(
      styles.base,
      {
        [styles.hasFocus]: value || hasFocus,
      },
      { [styles.hasError]: errorMessage },
      styles[theme],
      className,
    );
    const errorMessageId = `${name}-error-message`;

    const handleFocus = (event) => {
      if (onFocus) {
        onFocus(event);
      }

      setHasFocus(true);
    };

    const handleBlur = (event) => {
      if (onBlur) {
        onBlur(event);
      }

      setHasFocus(false);
    };

    return (
      <div className={classSet}>
        <label className={cx(styles.label)} htmlFor={name}>
          {label}
        </label>
        <select
          aria-describedby={errorMessageId}
          aria-invalid={!!errorMessage}
          aria-label={label || name}
          className={cx(styles.input, { [styles.isBlock]: isBlock })}
          data-test-ref={dataTestRef}
          id={name}
          name={name}
          onBlur={handleBlur}
          onChange={onChange}
          onFocus={handleFocus}
          ref={ref}
          value={value}
        >
          {label && <option value="" />}
          {options.map(({ id, label: optionLabel, value: optionValue }) => (
            <option key={id || optionValue} value={optionValue}>
              {optionLabel}
            </option>
          ))}
        </select>
        <Icon
          className={styles.icon}
          height={15}
          name="chevron"
          theme={theme}
          width={15}
        />
        {errorMessage && (
          <span className={styles.errorMessage} id={errorMessageId}>
            {errorMessage}
          </span>
        )}
      </div>
    );
  },
);

Select.propTypes = {
  className: PropTypes.string,
  dataTestRef: PropTypes.string,
  errorMessage: PropTypes.string,
  isBlock: PropTypes.bool,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func.isRequired,
  onFocus: PropTypes.func,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      label: PropTypes.string,
      value: PropTypes.string,
    }),
  ),
  theme: PropTypes.oneOf(['dark', 'light']),
  value: PropTypes.string,
};

Select.defaultProps = {
  className: undefined,
  dataTestRef: undefined,
  errorMessage: '',
  isBlock: false,
  label: undefined,
  name: undefined,
  onBlur: undefined,
  onChange: undefined,
  onFocus: undefined,
  options: [],
  theme: 'dark',
  value: undefined,
};

export { Select };
