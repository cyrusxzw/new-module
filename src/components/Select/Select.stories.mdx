import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { StorybookGridOverlay } from '~.storybook/components';
import { State } from '~.storybook/State';
import { Select } from './Select';
import { SelectFixture } from './Select.fixture';

<Meta title="Components/Select" component={Select} />

# Select

<Canvas>
  <Story name="Base component">
    {(() => {
      const implementation = {
        gridOverlay: {
          isActive: knobs.boolean('Show grid overlay', false, 'Implementation'),
          hasInvertedColours: knobs.boolean(
            'Invert grid overlay colours',
            false,
            'Implementation',
          ),
        },
      };
      return (
        <>
          <State
           initialState={{
             value: '',
           }}
          >
           {(state, setState) => {
             const errorMessage = knobs.text('errorMessage', '', 'Copy');
             const isBlock = knobs.boolean('isBlock', false, 'Presentation');
             const label = knobs.text('label', SelectFixture.label, 'Copy');
             const name = knobs.text('name', SelectFixture.name, 'Copy');
             const onChange = (event) => {
               event.persist();
               setState({...state, value: event.target.value})
             };
             const options = knobs.object('options', SelectFixture.options, 'Copy');
             const theme = knobs.select(
               'theme',
               ['dark', 'light'],
               'dark',
               'Presentation',
             );
             return (
               <Select
                 errorMessage={errorMessage}
                 isBlock={isBlock}
                 label={label}
                 name={name}
                 onChange={onChange}
                 options={options}
                 testReference={SelectFixture.testReference}
                 theme={theme}
                 value={state.value}
               />
             )
           }}
          </State>
          <StorybookGridOverlay
            isActive={implementation.gridOverlay.isActive}
            hasInvertedColours={implementation.gridOverlay.hasInvertedColours}
          />
        </>
      );
    })()}
  </Story>
</Canvas>

## API

<ArgsTable of={Select} />
