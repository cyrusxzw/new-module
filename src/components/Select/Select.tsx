import React, { forwardRef, useState } from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import { Icon } from '~/components/Icon';
import type { SelectProps } from './Select.types';
import styles from './Select.module.css';

const Select = forwardRef<HTMLSelectElement, SelectProps>(function SelectRef(
  {
    className,
    dataTestRef,
    errorMessage,
    isBlock = false,
    label,
    name,
    onBlur,
    onChange,
    onFocus,
    options,
    theme,
    value,
  },
  ref,
) {
  const currentTheme = useThemeContext(theme, 'dark');
  const [hasFocus, setHasFocus] = useState(false);
  const errorMessageId = `${name}-error-message`;

  const classSet = cx(
    styles.base,
    { [styles.isBlock]: isBlock },
    {
      [styles.hasFocus]: value || hasFocus,
    },
    { [styles.hasError]: errorMessage },
    styles[currentTheme],
    className,
  );

  const handleFocus = (event: React.FocusEvent<HTMLSelectElement>) => {
    onFocus?.(event);

    setHasFocus(true);
  };

  const handleBlur = (event: React.FocusEvent<HTMLSelectElement>) => {
    onBlur?.(event);

    setHasFocus(false);
  };

  return (
    <div className={classSet}>
      <label className={cx(styles.label)} htmlFor={name}>
        {label}
      </label>
      <select
        aria-describedby={errorMessage ? errorMessageId : undefined}
        aria-invalid={!!errorMessage}
        aria-label={label || name}
        className={styles.input}
        data-test-ref={dataTestRef}
        id={name}
        name={name}
        onBlur={handleBlur}
        onChange={onChange}
        onFocus={handleFocus}
        ref={ref}
        value={value}
      >
        {label && <option value="" />}
        {options.map(({ id, label: optionLabel, value: optionValue }) => (
          <option key={id || optionValue} value={optionValue}>
            {optionLabel}
          </option>
        ))}
      </select>
      <Icon
        className={styles.icon}
        height={15}
        name="chevron"
        theme={currentTheme}
        width={15}
      />
      {errorMessage && (
        <span className={styles.errorMessage} id={errorMessageId}>
          {errorMessage}
        </span>
      )}
    </div>
  );
});

export { Select };
