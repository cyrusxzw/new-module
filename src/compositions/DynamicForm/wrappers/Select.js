/* eslint-disable react/prop-types */
import React from 'react';
import { Controller } from 'react-hook-form/dist/index.ie11';
import { Select } from '~/components/Select';

export const SelectWrapper = ({
  className,
  control,
  dataTestRef,
  defaultValue,
  errorMessage,
  id,
  label,
  name,
  options,
  rules,
  theme,
}) => (
  <Controller
    control={control}
    defaultValue={defaultValue || ''}
    name={name}
    render={({ onChange, value }) => (
      <Select
        className={className}
        dataTestRef={dataTestRef}
        errorMessage={errorMessage}
        id={id}
        label={label}
        name={name}
        onChange={onChange}
        options={options}
        theme={theme}
        value={value}
      />
    )}
    rules={rules}
  />
);
