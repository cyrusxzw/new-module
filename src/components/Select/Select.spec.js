import React from 'react';
import { render } from '@testing-library/react';
import { Select } from './Select';
import { SelectFixture } from './Select.fixture';

describe('<Select />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <Select
        className={SelectFixture.className}
        errorMessage={SelectFixture.errorMessage}
        label={SelectFixture.label}
        name={SelectFixture.name}
        onBlur={SelectFixture.onBlur}
        onChange={SelectFixture.onChange}
        onFocus={SelectFixture.onFocus}
        options={SelectFixture.options}
        testReference={SelectFixture.testReference}
        value={SelectFixture.value}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
