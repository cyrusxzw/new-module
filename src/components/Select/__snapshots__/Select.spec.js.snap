// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render base component correctly 1`] = `
<div>
  <div
    class="base hasFocus hasError dark"
  >
    <label
      class="label"
      for="select"
    >
      Vestibulum accumsan
    </label>
    <select
      aria-describedby="select-error-message"
      aria-invalid="true"
      aria-label="Vestibulum accumsan"
      class="input"
      id="select"
      name="select"
    >
      <option
        value=""
      />
      <option
        value="morbi-id-nulla"
      >
        Morbi id nulla
      </option>
      <option
        value="nullam-ultrices"
      >
        Nullam ultrices
      </option>
    </select>
    <svg
      aria-labelledby="chevron-1"
      class="base chevron dark icon"
      data-testid="data-testid-Icon"
      focusable="false"
      height="15"
      role="img"
      style="height: 15px; width: 15px;"
      tabindex="-1"
      viewBox="0 0 50 50"
      width="15"
    >
      <title
        id="chevron-1"
      >
        chevron
      </title>
      <g>
        <polygon
          class="chevron"
          points="25,31.3 4.2,10.5 0.1,14.6 25,39.5 50,14.6 45.9,10.5"
        />
      </g>
    </svg>
    <span
      class="errorMessage"
      id="select-error-message"
    >
      Test Error Messsage
    </span>
  </div>
</div>
`;
